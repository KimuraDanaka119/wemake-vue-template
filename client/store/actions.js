import * as types from './types'

export default {
  incrementAsync ({ commit }) {
    setTimeout(() => {
      commit(types.INCREMENT)
    }, 200)
  }
}
