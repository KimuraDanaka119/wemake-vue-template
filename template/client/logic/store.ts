import { Module } from 'vuex-simple'

// TODO: document

import CommentsModule from '~/logic/comments/module'

export default class TypedStore {
  @Module()
  public comments = new CommentsModule()
}
