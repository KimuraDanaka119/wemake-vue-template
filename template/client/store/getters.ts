import { StateType } from '~/types'

const getters = {
  hasComments (state: StateType): boolean {
    return Boolean(state.comments && state.comments.length > 0)
  },
}

export default getters
