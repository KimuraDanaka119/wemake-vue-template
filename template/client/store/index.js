// @flow

import Vue from 'vue'

import comments from '~/logics/api/comments'
import * as mutationTypes from '~/types/mutations'
import type {
  CommentType,
  CommentPayloadType,
  StateType,
  RawCommentType
} from '~/types'

function state (): StateType {
  return {
    comments: null
  }
}

const getters = {
  hasComments (state: StateType): boolean {
    return Boolean(state.comments && state.comments.length > 0)
  }
}

const mutations = {
  [mutationTypes.SET_COMMENTS]: (
    state: StateType, comments: Array<RawCommentType>
  ) => {
    const updatedComments: Array<CommentType> = []

    for (const comment of comments.slice(0, 10)) {
      const newOne = {
        ...comment,
        rating: 0
      }
      updatedComments.push(newOne)
    }

    state.comments = updatedComments
  },

  [mutationTypes.UPDATE_RATING]: (
    state: StateType, { commentId, delta }: CommentPayloadType
  ) => {
    if (!state.comments) return

    const commentIndex = state.comments.findIndex((c: CommentType) => {
      return c.id === commentId
    })

    if (!state.comments || !state.comments[commentIndex]) return

    state.comments[commentIndex].rating += delta
  }
}

const actions = {
  async fetchComments ({ commit }: any, app: Vue) {
    const data = await comments.fetchComments(app)
    commit(mutationTypes.SET_COMMENTS, data)

    return data
  }
}

export {
  state,
  getters,
  mutations,
  actions
}
