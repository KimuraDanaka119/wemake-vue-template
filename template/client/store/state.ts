import { StateType } from '~/types'

function createState (): StateType {
  return {
    'comments': undefined,
  }
}

export default createState
