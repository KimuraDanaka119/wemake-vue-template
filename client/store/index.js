import Vue from 'vue'
import Vuex from 'vuex'
import createState from './state'
import mutations from './mutations'
import actions from './actions'

Vue.use(Vuex)

export default function () {
  return new Vuex.Store({
    state: createState(),
    mutations,
    actions
  })
}
