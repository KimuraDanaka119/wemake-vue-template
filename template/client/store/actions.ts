import { ActionContext, Store } from 'vuex'

import comments from '~/logics/api/comments'
import * as mutationTypes from '~/store/types'
import { StateType } from '~/types'

const actions = {
  async fetchComments (
    this: Store<StateType>,
    { commit }: ActionContext<StateType, StateType>,
  ) {
    const commentsList = await comments.fetchComments(this.$axios)
    commit(mutationTypes.SET_COMMENTS, commentsList)
    return commentsList
  },
}

export default actions
