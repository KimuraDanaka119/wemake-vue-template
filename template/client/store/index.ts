// We use default Nuxt Module-based store,
// read more about it here:
// https://nuxtjs.org/guide/vuex-store

// TODO: document

import Vue from 'vue'
import Vuex, { Store, Plugin } from 'vuex'
import { createVuexStore } from 'vuex-simple'
import { Container } from 'vue-typedi'

import tokens from '~/logic/tokens'
import TypedStore from '~/logic/store'
import { StateType } from '~/logic/types'

Vue.use(Vuex)

export default function store (
  ssrContext,
  { plugins = [] }: { plugins: Plugin<StateType>[] },
): Store<StateType> {
  if (!Container.has(tokens.AXIOS)) {
    Container.set(tokens.AXIOS, (): void => {})
  }

  const typedStore = new TypedStore()

  // Registering DI container items:
  Container.set(tokens.STORE, typedStore)

  return createVuexStore(typedStore, {
    plugins,
  })
}
