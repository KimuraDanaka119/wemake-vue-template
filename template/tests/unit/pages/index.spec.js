import Vuex from 'vuex'
import { mount, createLocalVue } from '@vue/test-utils'
import MockAdapter from 'axios-mock-adapter'

import Index from '~/pages/index'
import { actions, getters, state, mutations } from '~/store'

import { createStore } from '../../utils/store'

const localVue = createLocalVue()
localVue.use(Vuex)

const comments = [{
  id: 1,
  email: 'test@mail.com',
  body: 'some text',
  rating: -4
}, {
  id: 2,
  email: 'other@mail.com',
  body: 'other text',
  rating: 2
}]

const mockedComment = {
  id: 3,
  email: 'mock@mail.com',
  body: 'mock text',
  rating: 19
}

const propsData = { comments }

describe('unit tests for Index page', () => {
  let store

  beforeEach(() => {
    store = createStore({ actions, getters, state, mutations }, {
      comments
    })
  })

  test('should have two comments', () => {
    const wrapper = mount(Index, { store, localVue, propsData })
    expect(wrapper.findAll('.comment__component')).toHaveLength(2)
  })

  test('should load new comments on actions', async () => {
    const wrapper = mount(Index, { store, localVue, propsData })

    const mock = new MockAdapter(store.$axios)
    mock.onGet('/comments').reply(200, [mockedComment])

    await wrapper.vm.$store.dispatch('fetchComments', wrapper.vm)
    expect(wrapper.vm.$store.state.comments).toHaveLength(1)
    expect(wrapper.vm.$store.state.comments[0].email).toBe(mockedComment.email)
    expect(wrapper.findAll('.comment__component')).toHaveLength(1)
  })

  test('should match the snapshot', () => {
    const wrapper = mount(Index, { store, localVue, propsData })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
