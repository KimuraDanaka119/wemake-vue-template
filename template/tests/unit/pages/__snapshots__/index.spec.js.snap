// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit tests for Index page should match the snapshot 1`] = `
"<section><div><div class=\\"logo\\"><h1><a href=\\"https://github.com/wemake-services/wemake-vue-template\\">
      wemake-vue-template
    </a></h1> <h2><a href=\\"https://wemake-services.gitbook.io/wemake-vue-template/\\">
      documentation
    </a></h2></div> <div class=\\"actions\\"><button class=\\"reload\\">
        reload comments
      </button></div> <section class=\\"container\\"><div class=\\"commentComponent\\"><p class=\\"author\\">test@mail.com</p> <p class=\\"body\\">some text</p> <div class=\\"rating\\"><button class=\\"ratingUp\\">
      +
    </button> <span class=\\"number\\">-4</span> <button class=\\"ratingDown\\">
      -
    </button></div></div><div class=\\"commentComponent\\"><p class=\\"author\\">other@mail.com</p> <p class=\\"body\\">other text</p> <div class=\\"rating\\"><button class=\\"ratingUp\\">
      +
    </button> <span class=\\"number\\">2</span> <button class=\\"ratingDown\\">
      -
    </button></div></div></section></div></section>"
`;
